File: RemoveNthNodeEnd.java
// Time Complexity :O(n)
// Space Complexity :O(1)
// Did this code successfully run on Leetcode :yes
// Any problem you faced while coding this :no


//using slow and fast pointer and a dummy node for removing head node in a testcase
class Solution {
        public ListNode removeNthFromEnd(ListNode head, int n) {
            //dummy
            ListNode dummy = new ListNode(-1);
            dummy.next= head;


            ListNode fast = dummy;
            int count=0;
            
            //fast reaches nth node
            while(count<=n){
                fast=fast.next;
                count++;
            }
    
            ListNode slow = dummy;
            while(fast!=null){
                slow=slow.next;
                fast=fast.next;
            }
            //removing the node
            ListNode temp=slow.next;
            slow.next=slow.next.next;
            temp.next=null;

            //returning head
            return dummy.next;
            
        }
    }
    

File: Remove_nth_node_end.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        dup=ListNode(-1,-1)
        dup.next=head
        fast=dup
        slow=dup
        c=0
        while c<=n and fast!=None:
            fast=fast.next
            c+=1
            
        print(dup)
        # print(fast.val)
        while fast!=None:
            fast=fast.next
            slow=slow.next
        if slow.next!=None and slow.next.next!=None:
            slow.next=slow.next.next
        else:
            slow.next=None

        return dup.next
        